src/sources/logplex.rs

```diff
@@ -0,0 +1,190 @@
+use crate::{
+    event::{self, Event},
+    topology::config::{DataType, GlobalOptions, SourceConfig},
+};
+use bytes::Buf;
+use chrono::{DateTime, Utc};
+use futures::{sync::mpsc, Future, Sink};
+use serde::{Deserialize, Serialize};
+use std::sync::{Arc, Mutex};
+use std::{
+    io::{BufRead, BufReader},
+    net::SocketAddr,
+};
+use stream_cancel::Tripwire;
+use warp::Filter;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct LogplexConfig {
+    address: SocketAddr,
+}
+
+#[typetag::serde(name = "logplex")]
+impl SourceConfig for LogplexConfig {
+    fn build(
+        &self,
+        _: &str,
+        _: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let (trigger, tripwire) = Tripwire::new();
+        let trigger = Arc::new(Mutex::new(Some(trigger)));
+
+        let svc = warp::post2()
+            .and(warp::path("events"))
+            .and(warp::header::<usize>("Logplex-Msg-Count"))
+            .and(warp::header::<String>("Logplex-Frame-Id"))
+            .and(warp::header::<String>("Logplex-Drain-Token"))
+            .and(warp::body::concat())
+            .and_then(move |msg_count, frame_id, drain_token, body| {
+                info!(message = "Handling logplex request", %msg_count, %frame_id, %drain_token);
+
+                let events = body_to_events(body);
+
+                if events.len() != msg_count {
+                    error!(message = "Parsed event count does not match message count header", event_count = events.len(), %msg_count);
+                    debug_assert!(false, "Parsed event count does not match message count header");
+                }
+
+                let out = out.clone();
+                let trigger = trigger.clone();
+                out.send_all(futures::stream::iter_ok(events))
+                    .map_err(move |_: mpsc::SendError<Event>| {
+                        error!("Failed to forward events, downstream is closed");
+                        // shut down the http server if someone hasn't already
+                        trigger.try_lock().ok().take().map(drop);
+                        warp::reject::custom("shutting down")
+                })
+                .map(|_| warp::reply())
+            });
+
+        let ping = warp::get2().and(warp::path("ping")).map(|| "pong");
```

Comment:
is this required or was this just for testing?

Reply:
Not required, but it's convenient when setting up infrastructure to make sure the server is reachable. Think of it as a minimum viable healthcheck, since we have no other routes without side effects.